src/granite_orm/transactions.cr

```diff
@@ -14,27 +14,35 @@ module Granite::ORM::Transactions
 
       begin
         __run_before_save
-        if value = @{{primary_name}}
+        now = Time.now.to_utc
+
+        if (value = @{{primary_name}}) && !new_record?
           __run_before_update
-          @updated_at = Time.now.to_utc
+          @updated_at = now
           params_and_pk = params
           params_and_pk << value
           @@adapter.update @@table_name, @@primary_name, self.class.fields, params_and_pk
           __run_after_update
         else
           __run_before_create
-          @created_at = Time.now.to_utc
-          @updated_at = Time.now.to_utc
+          @created_at = @updated_at = now
+          params = params()
+          fields = self.class.fields
+          if value = @{{primary_name}}
+            fields << "{{primary_name}}"
+            params << value
+          end
           {% if primary_type.id == "Int32" %}
-            @{{primary_name}} = @@adapter.insert(@@table_name, self.class.fields, params).to_i32
+            @{{primary_name}} = @@adapter.insert(@@table_name, fields, params).to_i32
           {% else %}
-            @{{primary_name}} = @@adapter.insert(@@table_name, self.class.fields, params)
+            @{{primary_name}} = @@adapter.insert(@@table_name, fields, params)
           {% end %}
           __run_after_create
         end
+        @new_record = false
         __run_after_save
         return true
-      rescue ex : DB::Error
+      rescue ex
```

Comment:
Here, I deleted the restriction of `DB::Error` because saving with specified primary key causes various types of errors dependent on the adapter like this.
```crystal
raise "Duplicate entry '5' for key 'PRIMARY'"
```
I would like to know whether there is no compatibility problem.

Reply:
Thanks for nice advises! Unfortunately, the adapters throws **untyped** one. We need the other choices.